goamiddleware/jwt_middleware.go

```diff
@@ -0,0 +1,284 @@
+					scopesInClaim[scope] = true
+					scopesInClaimList = append(scopesInClaimList, scope)
+				}
+			case []interface{}:
+				for _, scope := range scopes {
+					if val, ok := scope.(string); ok {
+						scopesInClaim[val] = true
+						scopesInClaimList = append(scopesInClaimList, val)
+					}
+				}
+			default:
+				return nil, nil, fmt.Errorf("unsupported scope format in incoming JWT claim, was type %T", scopes)
+			}
+			break
+		}
+	}
+	sort.Strings(scopesInClaimList)
+	return scopesInClaim, scopesInClaimList, nil
+}
+
+// ErrJWTError is the error returned by this middleware when any sort of validation or assertion
+// fails during processing.
+var ErrJWTError = goa.NewErrorClass("jwt_security_error", 401)
+
+type contextKey int
+
+const (
+	jwtKey contextKey = iota + 1
+)
+
+// partitionKeys sorts keys by their type.
+func partitionKeys(k interface{}) ([]*rsa.PublicKey, []*ecdsa.PublicKey, [][]byte) {
+	var (
+		rsaKeys   []*rsa.PublicKey
+		ecdsaKeys []*ecdsa.PublicKey
+		hmacKeys  [][]byte
+	)
+
+	switch typed := k.(type) {
+	case []byte:
+		hmacKeys = append(hmacKeys, typed)
+	case [][]byte:
+		hmacKeys = typed
+	case string:
+		hmacKeys = append(hmacKeys, []byte(typed))
+	case []string:
+		for _, s := range typed {
+			hmacKeys = append(hmacKeys, []byte(s))
+		}
+	case *rsa.PublicKey:
+		rsaKeys = append(rsaKeys, typed)
+	case []*rsa.PublicKey:
+		rsaKeys = typed
+	case *ecdsa.PublicKey:
+		ecdsaKeys = append(ecdsaKeys, typed)
+	case []*ecdsa.PublicKey:
+		ecdsaKeys = typed
+	}
+
+	return rsaKeys, ecdsaKeys, hmacKeys
+}
+
+func validateRSAKeys(rsaKeys []*rsa.PublicKey, algo, incomingToken string) (token *jwt.Token, err error) {
+	for _, pubkey := range rsaKeys {
+		token, err = jwt.Parse(incomingToken, func(token *jwt.Token) (interface{}, error) {
+			if !strings.HasPrefix(token.Method.Alg(), algo) {
+				return nil, ErrJWTError(fmt.Sprintf("Unexpected signing method: %v", token.Header["alg"]))
+			}
+			return pubkey, nil
+		})
+		if err == nil {
+			return
+		}
+	}
+	return
+}
+
+func validateECDSAKeys(ecdsaKeys []*ecdsa.PublicKey, algo, incomingToken string) (token *jwt.Token, err error) {
+	for _, pubkey := range ecdsaKeys {
+		token, err = jwt.Parse(incomingToken, func(token *jwt.Token) (interface{}, error) {
+			if !strings.HasPrefix(token.Method.Alg(), algo) {
+				return nil, ErrJWTError(fmt.Sprintf("Unexpected signing method: %v", token.Header["alg"]))
+			}
+			return pubkey, nil
+		})
+		if err == nil {
+			return
+		}
+	}
+	return
+}
+
+func validateHMACKeys(hmacKeys [][]byte, algo, incomingToken string) (token *jwt.Token, err error) {
+	for _, key := range hmacKeys {
+		token, err = jwt.Parse(incomingToken, func(token *jwt.Token) (interface{}, error) {
+			if !strings.HasPrefix(token.Method.Alg(), algo) {
+				return nil, ErrJWTError(fmt.Sprintf("Unexpected signing method: %v", token.Header["alg"]))
+			}
+			return key, nil
+		})
+		if err == nil {
+			return
+		}
+	}
+	return
+}
```

Comment:
Is it possible to use the versions of these functions in jwt_token_context.go? Or move any shared functions to a common utility file?

Reply:
Shared functions: parse/validate{RSA,ECDSA,HMAC}Keys, partitionKeys